### PR TITLE
Use latest ggml from upstream

### PR DESCRIPTION
**Pull request by MollySophia**
What's changed:

- Correctly compile and inference with latest ggml.
- Regenerated ```expected-logits-*.bin``` using old code base, with ```GGML_SILU_FP16``` turned off. (Upstream ggml uses FP32 for SiLU now, while the old one uses FP16 for SiLU by default, which causes some logits difference in tests.)
- Slightly refactored CMakeLists.txt according to llama.cpp, and added option to enable Metal backend.
- Changed the way of offloading layers to gpu using the new ggml backends and scheduler.
- Use completely unmodified ggml submodule.
- Disabled thread sanitizer. It's broken in llama.cpp too.

TODOs:

- [x] Test basic inference
- [x] Test gpu offloading with Apple Metal 
- [x] Test gpu offloading with CUDA (And possibily fix any problems)
- [ ] Test gpu offloading with other backends (e.g HIP, OpenBLAS) (May not be possible myself :P)
- [ ] Update docs and README.md

### CODE REVIEW

**Comment by MollySophia** [comment]
Update: offloading with CUDA doesn't work yet
I'm working on it

**Review by LaylBongers — APPROVED** [review]
